# Assignment 2, Danylo Sovgut

# Question 1

In [5]:
import pandas as pd

# Denver Facility Data
denver_data = {
    "Line": ["BSC", "CFP"],
    "Annual Capacity (units)": [8_000_000, 5_000_000],
    "Fixed Costs ($/year)": [12_000_000, 8_000_000],
    "Variable Costs ($/unit)": [15, 8],
    "CO2 Emissions (kg/unit)": [2, 1],
    "Automation Level (%)": [65, 75]
}

denver_df = pd.DataFrame(denver_data)
denver_df


Line  Annual Capacity (units)  Fixed Costs ($/year)  \
0  BSC                  8000000              12000000   
1  CFP                  5000000               8000000   

   Variable Costs ($/unit)  CO2 Emissions (kg/unit)  Automation Level (%)  
0                       15                        2                    65  
1                        8                        1                    75

In [6]:
# Regional Demand Data
regional_demand = {
    "Region": ["Northeast", "Southeast", "Midwest", "Southwest", "Northwest"],
    "BSC Demand (units)": [2_500_000, 2_000_000, 1_800_000, 1_200_000, 1_000_000],
    "CFP Demand (units)": [1_500_000, 1_200_000, 1_000_000, 800_000, 500_000]
}

regional_demand_df = pd.DataFrame(regional_demand)
regional_demand_df


Region  BSC Demand (units)  CFP Demand (units)
0  Northeast             2500000             1500000
1  Southeast             2000000             1200000
2    Midwest             1800000             1000000
3  Southwest             1200000              800000
4  Northwest             1000000              500000

In [7]:
# Transportation and Emission Data
transport_emission_data = {
    "From_To": ["Denver_Northeast", "Denver_Southeast", "Denver_Midwest", "Denver_Southwest", "Denver_Northwest"],
    "Transport Cost ($/unit)": [8.5, 7.8, 5.2, 4.6, 4.2],
    "CO2 Emissions (kg/unit)": [5.2, 4.8, 3.1, 2.8, 2.5]
}

transport_emission_df = pd.DataFrame(transport_emission_data)
transport_emission_df


From_To  Transport Cost ($/unit)  CO2 Emissions (kg/unit)
0  Denver_Northeast                      8.5                      5.2
1  Denver_Southeast                      7.8                      4.8
2    Denver_Midwest                      5.2                      3.1
3  Denver_Southwest                      4.6                      2.8
4  Denver_Northwest                      4.2                      2.5

## a) Calculate total costs and emissions of serving all regions from Denver

In [54]:
from IPython.display import display, Markdown

# LaTeX formulas for cost and emissions
latex_formula = r"""
### Total Cost Calculation

1. **BSC Total Cost**:
$
\text{BSC Transport Cost} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Cost}_i)
$

$
\text{BSC Production Cost} = \text{BSC Variable Cost} \times \text{Total Demand for BSC}
$

$
\text{BSC Total Cost} = \text{BSC Fixed Cost} + \text{BSC Transport Cost} + \text{BSC Production Cost}
$

2. **CFP Total Cost**:
$
\text{CFP Transport Cost} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Cost}_i)
$

$
\text{CFP Production Cost} = \text{CFP Variable Cost} \times \text{Total Demand for CFP}
$

$
\text{CFP Total Cost} = \text{CFP Fixed Cost} + \text{CFP Transport Cost} + \text{CFP Production Cost}
$

3. **Total Cost for Both Items**:
$
\text{Total Cost} = \text{BSC Total Cost} + \text{CFP Total Cost}
$

---

### Total Emissions Calculation

1. **BSC Total Emissions**:
$
\text{BSC Transport Emissions} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Emissions}_i)
$

$
\text{BSC Production Emissions} = \text{BSC Emissions per Unit} \times \text{Total Demand for BSC}
$

$
\text{BSC Total Emissions} = \text{BSC Transport Emissions} + \text{BSC Production Emissions}
$

2. **CFP Total Emissions**:
$
\text{CFP Transport Emissions} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Emissions}_i)
$

$
\text{CFP Production Emissions} = \text{CFP Emissions per Unit} \times \text{Total Demand for CFP}
$

$
\text{CFP Total Emissions} = \text{CFP Transport Emissions} + \text{CFP Production Emissions}
$

3. **Total Emissions for Both Items**:
$
\text{Total Emissions} = \text{BSC Total Emissions} + \text{CFP Total Emissions}
$
"""

# Explanation of the process
explanation = r"""
**Process and Assumptions for Calculating Costs and Emissions:**

1. **Demand Distribution**:
   - The demand for both BSC and CFP items was distributed across five regions:
     - **Northeast, Southeast, Midwest, Southwest, Northwest.**

2. **Costs**:
   - Transportation costs were calculated by multiplying demand for each region by the transportation cost per item for that region.
   - Production costs were calculated by multiplying the total demand for each item by the respective variable cost per unit.
   - Total costs included fixed costs, transportation costs, and production costs.

3. **Emissions**:
   - Transportation emissions were calculated by multiplying demand for each region by the transport emissions per item for that region.
   - Production emissions were calculated by multiplying the total demand for each item by its emissions per unit.
   - Total emissions included both transportation and production emissions.

4. **Assumptions**:
   - All demand is met without exceeding facility capacity.
   - Costs and emissions scale linearly with demand and distance.
   - The selected cities accurately represent the regions geographically and economically.
   - The transportation mode (e.g., trucking) is consistent across all locations and distances.

"""

# Display the formulas and explanation
display(Markdown(latex_formula))
display(Markdown(explanation))



### Total Cost Calculation

1. **BSC Total Cost**:
$
\text{BSC Transport Cost} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Cost}_i)
$

$
\text{BSC Production Cost} = \text{BSC Variable Cost} \times \text{Total Demand for BSC}
$

$
\text{BSC Total Cost} = \text{BSC Fixed Cost} + \text{BSC Transport Cost} + \text{BSC Production Cost}
$

2. **CFP Total Cost**:
$
\text{CFP Transport Cost} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Cost}_i)
$

$
\text{CFP Production Cost} = \text{CFP Variable Cost} \times \text{Total Demand for CFP}
$

$
\text{CFP Total Cost} = \text{CFP Fixed Cost} + \text{CFP Transport Cost} + \text{CFP Production Cost}
$

3. **Total Cost for Both Items**:
$
\text{Total Cost} = \text{BSC Total Cost} + \text{CFP Total Cost}
$

---

### Total Emissions Calculation

1. **BSC Total Emissions**:
$
\text{BSC Transport Emissions} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Emissions}_i)
$

$
\text{BSC Production Emissions} = \text{BSC Emissions per Unit} \times \text{Total Demand for BSC}
$

$
\text{BSC Total Emissions} = \text{BSC Transport Emissions} + \text{BSC Production Emissions}
$

2. **CFP Total Emissions**:
$
\text{CFP Transport Emissions} = \sum_{i=1}^5 (\text{Demand}_i \times \text{Transport Emissions}_i)
$

$
\text{CFP Production Emissions} = \text{CFP Emissions per Unit} \times \text{Total Demand for CFP}
$

$
\text{CFP Total Emissions} = \text{CFP Transport Emissions} + \text{CFP Production Emissions}
$

3. **Total Emissions for Both Items**:
$
\text{Total Emissions} = \text{BSC Total Emissions} + \text{CFP Total Emissions}
$



**Process and Assumptions for Calculating Costs and Emissions:**

1. **Demand Distribution**:
   - The demand for both BSC and CFP items was distributed across five regions:
     - **Northeast, Southeast, Midwest, Southwest, Northwest.**

2. **Costs**:
   - Transportation costs were calculated by multiplying demand for each region by the transportation cost per item for that region.
   - Production costs were calculated by multiplying the total demand for each item by the respective variable cost per unit.
   - Total costs included fixed costs, transportation costs, and production costs.

3. **Emissions**:
   - Transportation emissions were calculated by multiplying demand for each region by the transport emissions per item for that region.
   - Production emissions were calculated by multiplying the total demand for each item by its emissions per unit.
   - Total emissions included both transportation and production emissions.

4. **Assumptions**:
   - All demand is met without exceeding facility capacity.
   - Costs and emissions scale linearly with demand and distance.
   - The selected cities accurately represent the regions geographically and economically.
   - The transportation mode (e.g., trucking) is consistent across all locations and distances.



In [15]:
import numpy as np

# Fixed costs
bsc_fixed_cost = 12_000_000  # BSC Fixed Cost
cfp_fixed_cost = 8_000_000   # CFP Fixed Cost

# Variable costs
bsc_variable_cost = 15  # BSC Variable Cost per unit
cfp_variable_cost = 8   # CFP Variable Cost per unit

# Demand array (1x5)
demand_array_bsc = np.array([2_500_000, 2_000_000, 1_800_000, 1_200_000, 1_000_000])
demand_array_cfp = np.array([1_500_000, 1_200_000, 1_000_000, 800_000, 500_000])

# Transportation cost array (1x5)
transport_cost_array = np.array([8.5, 7.8, 5.2, 4.6, 4.2])

# BSC Total Cost
bsc_transport_cost = np.dot(demand_array_bsc, transport_cost_array)  # Matrix multiplication
bsc_production_cost = bsc_variable_cost * np.sum(demand_array_bsc)
bsc_total_cost = bsc_fixed_cost + bsc_production_cost + bsc_transport_cost

# CFP Total Cost
cfp_transport_cost = np.dot(demand_array_cfp, transport_cost_array)
cfp_production_cost = cfp_variable_cost * np.sum(demand_array_cfp)
cfp_total_cost = cfp_fixed_cost + cfp_production_cost + cfp_transport_cost

#total cost
total_cost = bsc_total_cost + cfp_total_cost

# Print Results
print("BSC Total Cost: $", bsc_total_cost)
print("CFP Total Cost: $", cfp_total_cost)
print("Total Cost of Both items: $", total_cost)

BSC Total Cost: $ 195430000.0
CFP Total Cost: $ 81090000.0
Total Cost of Both items: $ 276520000.0


In [17]:
# Emissions per unit for production
bsc_emission_per_unit = 2  # kg/unit for BSC
cfp_emission_per_unit = 1  # kg/unit for CFP

# Transportation emissions array (1x5)
transport_emission_array = np.array([5.2, 4.8, 3.1, 2.8, 2.5])

# BSC Total Emissions
bsc_production_emissions = bsc_emission_per_unit * np.sum(demand_array_bsc)
bsc_transport_emissions = np.dot(demand_array_bsc, transport_emission_array)
bsc_total_emissions = bsc_production_emissions + bsc_transport_emissions

# CFP Total Emissions
cfp_production_emissions = cfp_emission_per_unit * np.sum(demand_array_cfp)
cfp_transport_emissions = np.dot(demand_array_cfp, transport_emission_array)
cfp_total_emissions = cfp_production_emissions + cfp_transport_emissions

total_total_emissions = cfp_total_emissions + bsc_total_emissions


# Print Results
print("BSC Total CO2 Emissions:", bsc_total_emissions, "kg")
print("CFP Total CO2 Emissions:", cfp_total_emissions, "kg")
print("Total CO2 Emissions:", total_total_emissions, "kg")


BSC Total CO2 Emissions: 51040000.0 kg
CFP Total CO2 Emissions: 25150000.0 kg
Total CO2 Emissions: 76190000.0 kg


## b) Evaluate capacity utilization and identify if Denver alone can handle projected demand 

In [18]:
# Total demand for BSC and CFP
total_bsc_demand = np.sum(demand_array_bsc)
total_cfp_demand = np.sum(demand_array_cfp)

# Denver's annual capacity
denver_bsc_capacity = 8_000_000
denver_cfp_capacity = 5_000_000

# Capacity utilization for BSC and CFP
bsc_utilization = (total_bsc_demand / denver_bsc_capacity) * 100
cfp_utilization = (total_cfp_demand / denver_cfp_capacity) * 100

# Print results
print(f"BSC Capacity Utilization: {bsc_utilization:.2f}%")
print(f"CFP Capacity Utilization: {cfp_utilization:.2f}%")


BSC Capacity Utilization: 106.25%
CFP Capacity Utilization: 100.00%


In [19]:
# Check if demand exceeds capacity
bsc_exceeds_capacity = total_bsc_demand > denver_bsc_capacity
cfp_exceeds_capacity = total_cfp_demand > denver_cfp_capacity

if bsc_exceeds_capacity:
    print("Denver cannot handle the BSC demand.")
else:
    print("Denver can handle the BSC demand.")

if cfp_exceeds_capacity:
    print("Denver cannot handle the CFP demand.")
else:
    print("Denver can handle the CFP demand.")


Denver cannot handle the BSC demand.
Denver can handle the CFP demand.


## c) Assess regional service levels (e.g., can Denver meet 2-day delivery for the East Coast?)

In [20]:
# Approximate distances from Denver to regions (in miles)
distances = {
    "Northeast": 1700,  # Denver to New York City
    "Southeast": 1400,  # Denver to Atlanta
    "Midwest": 800,     # Denver to Chicago
    "Southwest": 600,   # Denver to Phoenix
    "Northwest": 1300   # Denver to Seattle
}

# Travel speeds (average in mph)
average_speed = 55

# Calculate travel times (in hours)
travel_times = {region: distance / average_speed for region, distance in distances.items()}
print("Travel Times (hours):", travel_times)


Travel Times (hours): {'Northeast': 30.90909090909091, 'Southeast': 25.454545454545453, 'Midwest': 14.545454545454545, 'Southwest': 10.909090909090908, 'Northwest': 23.636363636363637}


In [21]:
# Check if delivery can be made within 48 hours
service_level = {region: "Meets 2-day delivery" if time <= 48 else "Does NOT meet 2-day delivery" 
                 for region, time in travel_times.items()}

print("Service Level Assessment:")
for region, status in service_level.items():
    print(f"{region}: {status}")


Service Level Assessment:
Northeast: Meets 2-day delivery
Southeast: Meets 2-day delivery
Midwest: Meets 2-day delivery
Southwest: Meets 2-day delivery
Northwest: Meets 2-day delivery


Based on the calculated travel times, Denver’s ability to meet the 2-day delivery requirement (48 hours) for all regions is feasible with the use of relay trucking. The estimated travel times for the Northeast (30.9 hours), Southeast (25.5 hours), Midwest (14.5 hours), Southwest (10.9 hours), and Northwest (23.6 hours) are all within the 48-hour window. However, in practice, a single driver cannot sustain continuous driving due to federal Hours of Service (HOS) regulations, which limit drivers to 11 hours of driving per day. To address this, relay trucking can be implemented, where multiple drivers take turns handling the freight at designated relay points. This ensures continuous movement of goods without violating regulations or overburdening individual drivers, making 2-day delivery achievable for even distant regions like the Northeast.

## d) Identify operational bottlenecks

The greatest operational bottleneck in the current system appears to be Denver’s inability to meet the required capacity. Whether due to labor constraints, inadequate infrastructure, or a lack of automation, Denver is struggling to keep pace with demand, leading to delays and inefficiencies. This issue is compounded by significant shipping delays, which add to the overall lead time. While shipping times are largely dependent on external logistics partners, they still create a considerable constraint, especially if third-party services cannot scale effectively to meet peak demand. To address these bottlenecks, it would be crucial to increase capacity in Denver through better labor management, automation, and infrastructure improvements. Additionally, optimizing the shipping process by negotiating better terms with logistics partners, improving internal distribution networks, and diversifying shipping routes could help alleviate delays and enhance operational efficiency. Both the capacity limitations in Denver and shipping constraints must be addressed to streamline the entire process and ensure timely delivery.

# Question 2

### a) Cost Implications: Calculate total supply chain cost (facility fixed, variable production, transport).

#### Below is an explanation of how I recalculated transportation costs and CO2 emissions for each of the new facilities.

In [59]:
from IPython.display import display, Markdown

# Use single backslashes instead of double
latex_formula = r"""
$
\text{Cost per mile} = \frac{\text{Transportation cost from Denver}}{\text{Distance from Denver to each region}}
$

$
\text{New transportation cost} = \text{Cost per mile} \times \text{Distance from new location to each region}
$

$
\text{CO2 emissions per mile} = \frac{\text{CO2 emissions from Denver}}{\text{Distance from Denver to each region}}
$

$
\text{New CO2 emissions} = \text{CO2 emissions per mile} \times \text{Distance from new location to each region}
$
"""

display(Markdown(latex_formula))
# Explanation of process and assumptions
explanation = """
**Process and Assumptions:**

1. **Distances and Representative Cities:**
   - We used the following cities as representatives for the regions:
     - **Northeast**: New York City (1780 miles from Denver)
     - **Southeast**: Miami (2065 miles from Denver)
     - **Midwest**: Chicago (1002 miles from Denver)
     - **Southwest**: Dallas (794 miles from Denver)
     - **Northwest**: Seattle (1320 miles from Denver)
   - Distances from the new locations (Charlotte, Columbus, Phoenix) to these regions were used for the calculation.

2. **Transportation Cost Calculation:**
   - The cost per mile was calculated as:
     \\[
     \\text{Cost per mile} = \\frac{\\text{Cost from Denver to region}}{\\text{Distance from Denver to region}}
     \\]
   - Using the cost per mile, the transportation costs from new locations were calculated.

3. **CO2 Emissions Calculation:**
   - Similarly, CO2 emissions per mile were calculated as:
     \\[
     \\text{CO2 per mile} = \\frac{\\text{CO2 emissions from Denver}}{\\text{Distance from Denver to region}}
     \\]
   - These values were then multiplied by the distances from the new locations to compute new CO2 emissions.

4. **Assumptions:**
   - We assume the cost per item and CO2 emissions are linear with respect to distance.
   - The selected cities accurately represent the regions geographically and economically.
   - The transportation mode (e.g., trucking) is consistent across all locations and distances.
"""

# Display LaTeX and explanation in Markdown
display(Markdown(latex_formula))
display(Markdown(explanation))



$
\text{Cost per mile} = \frac{\text{Transportation cost from Denver}}{\text{Distance from Denver to each region}}
$

$
\text{New transportation cost} = \text{Cost per mile} \times \text{Distance from new location to each region}
$

$
\text{CO2 emissions per mile} = \frac{\text{CO2 emissions from Denver}}{\text{Distance from Denver to each region}}
$

$
\text{New CO2 emissions} = \text{CO2 emissions per mile} \times \text{Distance from new location to each region}
$



$
\text{Cost per mile} = \frac{\text{Transportation cost from Denver}}{\text{Distance from Denver to each region}}
$

$
\text{New transportation cost} = \text{Cost per mile} \times \text{Distance from new location to each region}
$

$
\text{CO2 emissions per mile} = \frac{\text{CO2 emissions from Denver}}{\text{Distance from Denver to each region}}
$

$
\text{New CO2 emissions} = \text{CO2 emissions per mile} \times \text{Distance from new location to each region}
$



**Process and Assumptions:**

1. **Distances and Representative Cities:**
   - We used the following cities as representatives for the regions:
     - **Northeast**: New York City (1780 miles from Denver)
     - **Southeast**: Miami (2065 miles from Denver)
     - **Midwest**: Chicago (1002 miles from Denver)
     - **Southwest**: Dallas (794 miles from Denver)
     - **Northwest**: Seattle (1320 miles from Denver)
   - Distances from the new locations (Charlotte, Columbus, Phoenix) to these regions were used for the calculation.

2. **Transportation Cost Calculation:**
   - The cost per mile was calculated as:
     \[
     \text{Cost per mile} = \frac{\text{Cost from Denver to region}}{\text{Distance from Denver to region}}
     \]
   - Using the cost per mile, the transportation costs from new locations were calculated.

3. **CO2 Emissions Calculation:**
   - Similarly, CO2 emissions per mile were calculated as:
     \[
     \text{CO2 per mile} = \frac{\text{CO2 emissions from Denver}}{\text{Distance from Denver to region}}
     \]
   - These values were then multiplied by the distances from the new locations to compute new CO2 emissions.

4. **Assumptions:**
   - We assume the cost per item and CO2 emissions are linear with respect to distance.
   - The selected cities accurately represent the regions geographically and economically.
   - The transportation mode (e.g., trucking) is consistent across all locations and distances.


In [42]:
import numpy as np
import pandas as pd

# Input Data
regions = ["Northeast", "Southeast", "Midwest", "Southwest", "Northwest"]
denver_distances = np.array([1780, 2065, 1002, 794, 1320])  # in miles
denver_transport_costs = np.array([8.5, 7.8, 5.2, 4.6, 4.2])  # $ per item
denver_co2 = np.array([5.2, 4.8, 3.1, 2.8, 2.5])  # kg CO2 per item

# Cost and CO2 per mile
cost_per_mile = denver_transport_costs / denver_distances
co2_per_mile = denver_co2 / denver_distances

# Distances from new locations to regions
location_distances = {
    "Charlotte": np.array([634, 725, 755, 1015, 2800]),
    "Columbus": np.array([536, 1030, 360, 1000, 2400]),
    "Phoenix": np.array([2400, 2375, 1750, 1065, 1430]),
}

# Compute transportation costs and CO2 emissions
transport_costs_array = []
co2_array = []

for location, distances in location_distances.items():
    costs = cost_per_mile * distances  # Transportation costs
    co2 = co2_per_mile * distances    # CO2 emissions
    transport_costs_array.append(costs)
    co2_array.append(co2)

# Convert results to numpy arrays
transport_costs_array = np.array(transport_costs_array)
co2_array = np.array(co2_array)

# Create DataFrames for display
cost_results = pd.DataFrame(
    transport_costs_array, 
    index=location_distances.keys(), 
    columns=[f"Cost to {region}" for region in regions]
)

co2_results = pd.DataFrame(
    co2_array, 
    index=location_distances.keys(), 
    columns=[f"CO2 to {region}" for region in regions]
)

In [109]:
co2_results

CO2 to Northeast  CO2 to Southeast  CO2 to Midwest  \
Charlotte          1.852135          1.685230        2.335828   
Columbus           1.565843          2.394189        1.113772   
Phoenix            7.011236          5.520581        5.414172   

           CO2 to Southwest  CO2 to Northwest  
Charlotte          3.579345          5.303030  
Columbus           3.526448          4.545455  
Phoenix            3.755668          2.708333

In [61]:
# Display Results
print("Updated Transportation Costs ($) to Charlotte:")
display(cost_results)

print("\nCO2 Emissions (kg) to Charlotte:")
display(co2_results)

Updated Transportation Costs ($) to Charlotte:


Cost to Northeast  Cost to Southeast  Cost to Midwest  \
Charlotte           3.027528           2.738499         3.918164   
Columbus            2.559551           3.890557         1.868263   
Phoenix            11.460674           8.970944         9.081836   

           Cost to Southwest  Cost to Northwest  
Charlotte           5.880353           8.909091  
Columbus            5.793451           7.636364  
Phoenix             6.170025           4.550000


CO2 Emissions (kg) to Charlotte:


CO2 to Northeast  CO2 to Southeast  CO2 to Midwest  \
Charlotte          1.852135          1.685230        2.335828   
Columbus           1.565843          2.394189        1.113772   
Phoenix            7.011236          5.520581        5.414172   

           CO2 to Southwest  CO2 to Northwest  
Charlotte          3.579345          5.303030  
Columbus           3.526448          4.545455  
Phoenix            3.755668          2.708333

In [62]:
print (transport_costs_array)

[[ 3.02752809  2.73849879  3.91816367  5.88035264  8.90909091]
 [ 2.55955056  3.8905569   1.86826347  5.79345088  7.63636364]
 [11.46067416  8.97094431  9.08183633  6.17002519  4.55      ]]


In [63]:
print (co2_array)

[[1.85213483 1.68523002 2.33582834 3.57934509 5.3030303 ]
 [1.5658427  2.39418886 1.11377246 3.52644836 4.54545455]
 [7.01123596 5.52058111 5.41417166 3.75566751 2.70833333]]


In [65]:
import numpy as np
import pandas as pd

# Input Data
transport_cost_array = np.array([
    [3.02752809, 2.73849879, 3.91816367, 5.88035264, 8.90909091],  # Charlotte
    [2.55955056, 3.8905569, 1.86826347, 5.79345088, 7.63636364],  # Columbus
    [11.46067416, 8.97094431, 9.08183633, 6.17002519, 4.55]        # Phoenix
])

co2_array = np.array([
    [1.85213483, 1.68523002, 2.33582834, 3.57934509, 5.3030303],  # Charlotte
    [1.5658427, 2.39418886, 1.11377246, 3.52644836, 4.54545455],  # Columbus
    [7.01123596, 5.52058111, 5.41417166, 3.75566751, 2.70833333]  # Phoenix
])

# Demand Arrays (in units)
demand_bsc = np.array([2500000, 2000000, 1800000, 1200000, 1000000])  # BSC demand
demand_cfp = np.array([1500000, 1200000, 1000000, 800000, 500000])    # CFP demand

# Facility Setup Costs
bsc_fixed_cost = 7000000  # $ per year
cfp_fixed_cost = 5000000  # $ per year

bsc_variable_cost = 16  # $ per unit
cfp_variable_cost = 8.5  # $ per unit

# Function to calculate total cost for a location
def calculate_total_cost(location_index):
    # Transportation costs
    bsc_transport_cost = np.dot(transport_cost_array[location_index], demand_bsc)
    cfp_transport_cost = np.dot(transport_cost_array[location_index], demand_cfp)

    # Production costs
    bsc_production_cost = bsc_variable_cost * np.sum(demand_bsc)
    cfp_production_cost = cfp_variable_cost * np.sum(demand_cfp)

    # Fixed costs
    bsc_total_fixed = bsc_fixed_cost
    cfp_total_fixed = cfp_fixed_cost

    # Total cost
    bsc_total_cost = bsc_total_fixed + bsc_production_cost + bsc_transport_cost
    cfp_total_cost = cfp_total_fixed + cfp_production_cost + cfp_transport_cost

    total_cost = bsc_total_cost + cfp_total_cost

    return bsc_total_cost, cfp_total_cost, total_cost

# Calculate costs for all locations
locations = ["Charlotte", "Columbus", "Phoenix"]
results = []

for i, location in enumerate(locations):
    bsc_cost, cfp_cost, total_cost = calculate_total_cost(i)
    results.append([location, bsc_cost, cfp_cost, total_cost])

# Create a DataFrame for output
columns = ["Location", "BSC Total Cost ($)", "CFP Total Cost ($)", "Total Cost ($)"]
results_df = pd.DataFrame(results, columns=columns)

# Display results
display(results_df)

# Save to CSV if needed
#results_df.to_csv("supply_chain_costs.csv", index=False)


Location  BSC Total Cost ($)  CFP Total Cost ($)  Total Cost ($)
0  Charlotte        1.790640e+08        6.840448e+07    2.474685e+08
1   Columbus        1.751314e+08        6.632920e+07    2.414606e+08
2    Phoenix        2.178949e+08        9.174900e+07    3.096439e+08

In [71]:
import numpy as np
import pandas as pd

# Input Data
transport_cost_array = np.array([
    [3.02752809, 2.73849879, 3.91816367, 5.88035264, 8.90909091],  # Charlotte
    [2.55955056, 3.8905569, 1.86826347, 5.79345088, 7.63636364],  # Columbus
    [11.46067416, 8.97094431, 9.08183633, 6.17002519, 4.55]        # Phoenix
])

co2_array = np.array([
    [1.85213483, 1.68523002, 2.33582834, 3.57934509, 5.3030303],  # Charlotte
    [1.5658427, 2.39418886, 1.11377246, 3.52644836, 4.54545455],  # Columbus
    [7.01123596, 5.52058111, 5.41417166, 3.75566751, 2.70833333]  # Phoenix
])

# Denver Data
denver_transport_cost_array = np.array([8.5, 7.8, 5.2, 4.6, 4.2])

demand_bsc = np.array([2500000, 2000000, 1800000, 1200000, 1000000])  # BSC demand
demand_cfp = np.array([1500000, 1200000, 1000000, 800000, 500000])    # CFP demand

# Facility Setup Costs
bsc_fixed_cost_new = 7000000  # New facilities fixed cost
cfp_fixed_cost_new = 5000000  # New facilities fixed cost

bsc_variable_cost_new = 16  # New facilities variable cost per unit
cfp_variable_cost_new = 8.5  # New facilities variable cost per unit

# Denver Setup Costs
bsc_fixed_cost_denver = 12000000
cfp_fixed_cost_denver = 8000000

bsc_variable_cost_denver = 15
cfp_variable_cost_denver = 8

# Function to calculate total cost for a location
def calculate_total_cost(location_index, is_denver=False):
    if is_denver:
        transport_cost = np.dot(denver_transport_cost_array, demand_bsc) + np.dot(denver_transport_cost_array, demand_cfp)
        production_cost = (bsc_variable_cost_denver * np.sum(demand_bsc)) + (cfp_variable_cost_denver * np.sum(demand_cfp))
        fixed_cost = bsc_fixed_cost_denver + cfp_fixed_cost_denver
    else:
        transport_cost = np.dot(transport_cost_array[location_index], demand_bsc) + np.dot(transport_cost_array[location_index], demand_cfp)
        production_cost = (bsc_variable_cost_new * np.sum(demand_bsc)) + (cfp_variable_cost_new * np.sum(demand_cfp))
        fixed_cost = bsc_fixed_cost_new + cfp_fixed_cost_new

    total_cost = transport_cost + production_cost + fixed_cost
    return total_cost / 1e6  # Convert to millions

# Calculate costs for all locations
locations = ["Charlotte", "Columbus", "Phoenix", "Denver"]
results = []

for i, location in enumerate(locations):
    if location == "Denver":
        total_cost = calculate_total_cost(i, is_denver=True)
    else:
        total_cost = calculate_total_cost(i)
    results.append([location, total_cost])

# Create a DataFrame for output
columns = ["Location", "Total Cost (Millions $)"]
results_df = pd.DataFrame(results, columns=columns)

# Format costs to 2 decimal places
results_df["Total Cost (Millions $)"] = results_df["Total Cost (Millions $)"].round(2)

# Display results
display(results_df)

# Save to CSV if needed
results_df.to_csv("supply_chain_costs_with_denver.csv", index=False)


Location  Total Cost (Millions $)
0  Charlotte                   247.47
1   Columbus                   241.46
2    Phoenix                   309.64
3     Denver                   276.52

## Above is the cost of simply adding another fascility instead of Denver which after doing this whole assignment, I have realized that I need to completea combination of Denver plus another fascility so hwere is the following approach:

To avoid exceeding any single facility’s production capacity, we divide the total demand between Denver and each potential new site **(Charlotte, Columbus, or Phoenix)**. For instance, in the “Denver + Charlotte” scenario, Charlotte handles all BSC and CFP demand for the Northeast plus a *portion* of the Southeast, while Denver manages the remaining Southeast along with the Midwest, Southwest, and Northwest. This partial assignment ensures **Charlotte’s** load stays below its 4 million BSC and 3 million CFP annual limits. Similarly, we create splits for “Denver + Columbus” and “Denver + Phoenix,” adjusting which regions (or partial regions) each site serves so that the new facility’s demand also remains within its respective capacities. These predefined, feasible allocations then allow us to calculate the total cost (production + transport + fixed costs) for each **two-facility** network scenario, and compare them to the baseline of “all from Denver.”

In [128]:
import numpy as np
import pandas as pd

# ============================================
# 1) DEMAND & COST DATA
# ============================================

# Demand by region index: 0=NE, 1=SE, 2=MW, 3=SW, 4=NW
demand_bsc = np.array([2_500_000, 2_000_000, 1_800_000, 1_200_000, 1_000_000])
demand_cfp = np.array([1_500_000, 1_200_000, 1_000_000,   800_000,   500_000])

# Denver details (transport, production, fixed, etc.)
denver_trans_cost    = np.array([8.5, 7.8, 5.2, 4.6, 4.2])  # $/unit to [NE,SE,MW,SW,NW]
denver_bsc_variable  = 15
denver_cfp_variable  = 8
denver_bsc_fixed     = 12_000_000
denver_cfp_fixed     = 8_000_000

# Potential new facility data
# We'll use simpler, truncated transport cost arrays:
facility_trans_costs = {
    "Charlotte": np.array([3.03, 2.74, 3.92, 5.88, 8.91]),
    "Columbus":  np.array([2.56, 3.89, 1.87, 5.79, 7.64]),
    "Phoenix":   np.array([11.46,8.97, 9.08, 6.17, 4.55])
}
new_bsc_variable = 16
new_cfp_variable = 8.5
new_bsc_fixed    = 7_000_000
new_cfp_fixed    = 5_000_000

# HARDCODED baseline cost for "all from Denver" scenario
baseline_cost_m = 276.52

# ============================================
# 2) SCENARIO SPLITS
#    We create partial splits that don't exceed
#    each new facility's capacity (4M BSC, 3M CFP).
# ============================================

scenarios = {
    # 1) Denver + Charlotte
    # Charlotte: NE + partial SE
    # Denver: leftover SE + MW + SW + NW
    "Denver + Charlotte": {
        "Charlotte": {
            "BSC": {0:2_500_000, 1:1_000_000},  # NE=2.5M, SE=1.0M
            "CFP": {0:1_500_000, 1: 600_000}
        },
        "Denver": {
            "BSC": {1:1_000_000, 2:1_800_000, 3:1_200_000, 4:1_000_000},
            "CFP": {1: 600_000,  2:1_000_000, 3: 800_000,  4: 500_000}
        }
    },

    # 2) Denver + Columbus
    # Columbus: NE + partial MW + partial SE
    # Denver: leftover MW + leftover SE + SW + NW
    "Denver + Columbus": {
        "Columbus": {
            "BSC": {0:2_500_000, 2:900_000, 1:600_000},
            "CFP": {0:1_500_000, 2:500_000, 1:300_000}
        },
        "Denver": {
            "BSC": {2:900_000, 1:1_400_000, 3:1_200_000, 4:1_000_000},
            "CFP": {2:500_000, 1: 900_000,  3: 800_000,  4: 500_000}
        }
    },

    # 3) Denver + Phoenix
    # Phoenix: entire SE + NW
    # Denver: NE + MW + SW
    "Denver + Phoenix": {
        "Phoenix": {
            "BSC": {1:2_000_000, 4:1_000_000},
            "CFP": {1:1_200_000, 4: 500_000}
        },
        "Denver": {
            "BSC": {0:2_500_000, 2:1_800_000, 3:1_200_000},
            "CFP": {0:1_500_000, 2:1_000_000, 3: 800_000}
        }
    }
}

# ============================================
# 3) HELPER FUNCTIONS
# ============================================

def calculate_new_fac_cost(fac_name, bsc_dict, cfp_dict):
    """
    For the 'new' facility (Charlotte, Columbus, or Phoenix):
    - Sum transport cost for BSC+CFP from assigned region splits
    - Sum production cost
    - Add fixed cost (for both BSC & CFP lines)
    """
    trans_arr = facility_trans_costs[fac_name]
    
    # Transport cost
    t_cost = 0
    for r, qty in bsc_dict.items():
        t_cost += trans_arr[r] * qty
    for r, qty in cfp_dict.items():
        t_cost += trans_arr[r] * qty
    
    # Production cost
    bsc_total = sum(bsc_dict.values())
    cfp_total = sum(cfp_dict.values())
    p_cost = bsc_total * new_bsc_variable + cfp_total * new_cfp_variable
    
    # Fixed cost (both lines)
    f_cost = new_bsc_fixed + new_cfp_fixed
    return t_cost + p_cost + f_cost

def calculate_denver_cost(bsc_dict, cfp_dict):
    """
    Denver's cost for its portion:
    - Transport cost
    - Production cost
    - Denver's (BSC + CFP) fixed cost
    """
    t_cost = 0
    for r, qty in bsc_dict.items():
        t_cost += denver_trans_cost[r] * qty
    for r, qty in cfp_dict.items():
        t_cost += denver_trans_cost[r] * qty
    
    bsc_total = sum(bsc_dict.values())
    cfp_total = sum(cfp_dict.values())
    p_cost = bsc_total * denver_bsc_variable + cfp_total * denver_cfp_variable
    
    # Always count the fixed cost for Denver (both lines)
    f_cost = denver_bsc_fixed + denver_cfp_fixed
    return t_cost + p_cost + f_cost

# ============================================
# 4) EVALUATE SCENARIOS
# ============================================
results = []
for scenario_name, alloc in scenarios.items():
    # The "new facility" is whichever key isn't "Denver"
    new_fac_name = [k for k in alloc.keys() if k != "Denver"][0]
    
    # Extract each facility's dict
    nf_bsc_dict = alloc[new_fac_name]["BSC"]
    nf_cfp_dict = alloc[new_fac_name]["CFP"]
    dv_bsc_dict = alloc["Denver"]["BSC"]
    dv_cfp_dict = alloc["Denver"]["CFP"]
    
    # Calculate total cost
    cost_new_fac = calculate_new_fac_cost(new_fac_name, nf_bsc_dict, nf_cfp_dict)
    cost_denver  = calculate_denver_cost(dv_bsc_dict, dv_cfp_dict)
    total_cost   = cost_new_fac + cost_denver
    
    # Convert to millions
    total_m = round(total_cost / 1e6, 2)
    # "Savings" = baseline (hardcoded 276.52) - scenario cost
    savings = round(276.52 - total_m, 2)
    
    results.append([scenario_name, total_m, savings])

df = pd.DataFrame(results, columns=["Scenario", "Total Cost (Millions $)", "Savings vs Denver (Millions $)"])
df



Scenario  Total Cost (Millions $)  Savings vs Denver (Millions $)
0  Denver + Charlotte                   263.09                           13.43
1   Denver + Columbus                   261.73                           14.79
2    Denver + Phoenix                   296.64                          -20.12

In [126]:
import numpy as np
import pandas as pd

# ---------------------------
# Demand (units) by region index:
# 0=NE, 1=SE, 2=MW, 3=SW, 4=NW
# ---------------------------
demand_bsc = np.array([2_500_000, 2_000_000, 1_800_000, 1_200_000, 1_000_000])
demand_cfp = np.array([1_500_000, 1_200_000, 1_000_000,   800_000,   500_000])

# Denver
denver_trans_cost = np.array([8.5, 7.8, 5.2, 4.6, 4.2])
denver_bsc_capacity = 8_000_000
denver_cfp_capacity = 5_000_000
denver_bsc_fixed     = 12_000_000
denver_cfp_fixed     = 8_000_000
denver_bsc_variable  = 15
denver_cfp_variable  = 8

# New facility data
facility_names     = ["Charlotte","Columbus","Phoenix"]
facility_trans_cost= {
    "Charlotte": np.array([3.03, 2.74, 3.92, 5.88, 8.91]),
    "Columbus":  np.array([2.56, 3.89, 1.87, 5.79, 7.64]),
    "Phoenix":   np.array([11.46,8.97, 9.08, 6.17, 4.55])
}
new_bsc_capacity   = 4_000_000
new_cfp_capacity   = 3_000_000
new_bsc_fixed      = 7_000_000
new_cfp_fixed      = 5_000_000
new_bsc_variable   = 16
new_cfp_variable   = 8.5

# 1) Baseline: all from Denver
def cost_all_from_denver():
    total_bsc = demand_bsc.sum()
    total_cfp = demand_cfp.sum()
    # Check capacity
    if total_bsc > denver_bsc_capacity or total_cfp > denver_cfp_capacity:
        return np.inf
    # Costs
    transport_cost = np.sum(denver_trans_cost * (demand_bsc + demand_cfp))
    production_cost = total_bsc*denver_bsc_variable + total_cfp*denver_cfp_variable
    fixed_cost = denver_bsc_fixed + denver_cfp_fixed
    return (transport_cost + production_cost + fixed_cost)/1e6

baseline_cost_m = cost_all_from_denver()

############################################################################
# 2) We'll define the partial splits in a dictionary
#    Each scenario is "Denver + <facility>", but we specify how many units
#    each of those two facilities produce for each region (BSC & CFP).
############################################################################

scenarios = {
    "Denver + Charlotte": {
        "Charlotte": {
            "BSC": {
                0: 2_500_000,  # NE full 2.5M
                1: 1_000_000   # partial SE: 1.0M out of 2.0M
            },
            "CFP": {
                0: 1_500_000,  # NE full 1.5M
                1: 600_000     # partial SE: 0.6M out of 1.2M
            }
        },
        "Denver": {
            "BSC": {
                1: 1_000_000,  # leftover SE
                2: 1_800_000,  # MW
                3: 1_200_000,  # SW
                4: 1_000_000   # NW
            },
            "CFP": {
                1: 600_000,    # leftover SE
                2: 1_000_000,  # MW
                3: 800_000,    # SW
                4: 500_000     # NW
            }
        }
    },

    "Denver + Columbus": {
        "Columbus": {
            "BSC": {
                0: 2_500_000, # NE
                2: 900_000,   # partial MW
                1: 600_000    # partial SE
            },
            "CFP": {
                0: 1_500_000, # NE
                2: 500_000,   # partial MW
                1: 300_000    # partial SE
            }
        },
        "Denver": {
            "BSC": {
                2: 900_000,   # leftover MW
                1: 1_400_000, # leftover SE
                3: 1_200_000, # SW
                4: 1_000_000  # NW
            },
            "CFP": {
                2: 500_000,   # leftover MW
                1: 900_000,   # leftover SE
                3: 800_000,   # SW
                4: 500_000    # NW
            }
        }
    },

    "Denver + Phoenix": {
        "Phoenix": {
            "BSC": {
                1: 2_000_000, # SE
                4: 1_000_000  # NW
            },
            "CFP": {
                1: 1_200_000, # SE
                4: 500_000    # NW
            }
        },
        "Denver": {
            "BSC": {
                0: 2_500_000, # NE
                2: 1_800_000, # MW
                3: 1_200_000  # SW
            },
            "CFP": {
                0: 1_500_000, # NE
                2: 1_000_000, # MW
                3: 800_000    # SW
            }
        }
    }
}

############################################################################
# 3) Function to compute total cost for a scenario
############################################################################
def calculate_cost_for_scenario(fac_name, facility_allocation):
    """
    fac_name: e.g. "Charlotte", "Columbus", or "Phoenix"
    facility_allocation: e.g. the dictionary of { "BSC": {region: units}, "CFP": {...} }
    """
    # 1) check capacity
    bsc_total = sum(facility_allocation["BSC"].values())
    cfp_total = sum(facility_allocation["CFP"].values())
    if bsc_total > new_bsc_capacity or cfp_total > new_cfp_capacity:
        return np.inf
    
    # 2) transport cost
    trans_arr = facility_trans_cost[fac_name]
    transport_cost = 0
    for region, bsc_qty in facility_allocation["BSC"].items():
        transport_cost += trans_arr[region] * bsc_qty
    for region, cfp_qty in facility_allocation["CFP"].items():
        transport_cost += trans_arr[region] * cfp_qty
    
    # 3) production cost
    prod_cost = (bsc_total * new_bsc_variable) + (cfp_total * new_cfp_variable)
    
    # 4) fixed cost (both lines open)
    fix_cost = new_bsc_fixed + new_cfp_fixed

    return transport_cost + prod_cost + fix_cost


def calculate_cost_denver(denver_allocation):
    # check capacity
    bsc_total = sum(denver_allocation["BSC"].values())
    cfp_total = sum(denver_allocation["CFP"].values())
    if bsc_total > denver_bsc_capacity or cfp_total > denver_cfp_capacity:
        return np.inf
    
    # transport cost
    transport_cost = 0
    for region, bsc_qty in denver_allocation["BSC"].items():
        transport_cost += denver_trans_cost[region] * bsc_qty
    for region, cfp_qty in denver_allocation["CFP"].items():
        transport_cost += denver_trans_cost[region] * cfp_qty
    
    # production cost
    prod_cost = (bsc_total * denver_bsc_variable) + (cfp_total * denver_cfp_variable)
    
    # fixed cost for Denver (both lines presumably active)
    fix_cost = denver_bsc_fixed + denver_cfp_fixed
    
    return transport_cost + prod_cost + fix_cost

############################################################################
# 4) Evaluate each scenario
############################################################################
rows = []
for scenario_name, allocations in scenarios.items():
    # allocations = { "Charlotte": {...}, "Denver": {...} }, etc.
    # figure out which new facility is in the dict
    new_fac = [k for k in allocations.keys() if k != "Denver"][0]
    
    cost_new_fac = calculate_cost_for_scenario(new_fac, allocations[new_fac])
    cost_denver  = calculate_cost_denver(allocations["Denver"])
    
    total_cost   = cost_new_fac + cost_denver
    cost_millions= total_cost/1e6

    if baseline_cost_m == np.inf:
        savings = np.nan
    else:
        savings = baseline_cost_m - cost_millions

    rows.append([scenario_name, cost_millions, savings])

df_scenarios = pd.DataFrame(rows, columns=["Scenario","Total Cost (Millions $)","Savings vs Denver Only (Millions $)"])
df_scenarios["Total Cost (Millions $)"] = df_scenarios["Total Cost (Millions $)"].round(2)
df_scenarios["Savings vs Denver Only (Millions $)"] = df_scenarios["Savings vs Denver Only (Millions $)"].round(2)

print(f"Baseline (All from Denver) = {baseline_cost_m:.2f} M\n")
display(df_scenarios)


Baseline (All from Denver) = inf M



Scenario  Total Cost (Millions $)  \
0  Denver + Charlotte                   263.09   
1   Denver + Columbus                   261.73   
2    Denver + Phoenix                   296.64   

   Savings vs Denver Only (Millions $)  
0                                  NaN  
1                                  NaN  
2                                  NaN

### b) Emission Impacts: Compare CO2 footprint vs. continuing to ship from Denver

In [68]:
import numpy as np
import pandas as pd

# Input Data
transport_co2_array = np.array([
    [1.85213483, 1.68523002, 2.33582834, 3.57934509, 5.3030303],  # Charlotte
    [1.5658427, 2.39418886, 1.11377246, 3.52644836, 4.54545455],  # Columbus
    [7.01123596, 5.52058111, 5.41417166, 3.75566751, 2.70833333]  # Phoenix
])

denver_co2_array = np.array([5.2, 4.8, 3.1, 2.8, 2.5])  # Denver CO2 emissions

# Demand Arrays (in units)
demand_bsc = np.array([2500000, 2000000, 1800000, 1200000, 1000000])  # BSC demand
demand_cfp = np.array([1500000, 1200000, 1000000, 800000, 500000])    # CFP demand

# CO2 Emissions per Unit for Production
bsc_emission_per_unit_new = 1.8  # kg/unit for BSC in new facilities
cfp_emission_per_unit_new = 0.9  # kg/unit for CFP in new facilities

bsc_emission_per_unit_denver = 2  # kg/unit for BSC in Denver
cfp_emission_per_unit_denver = 1  # kg/unit for CFP in Denver

# Function to calculate total CO2 emissions for a location
def calculate_total_emissions(location_index, is_denver=False):
    if is_denver:
        transport_emissions = np.dot(denver_co2_array, demand_bsc) + np.dot(denver_co2_array, demand_cfp)
        production_emissions = (bsc_emission_per_unit_denver * np.sum(demand_bsc)) + (cfp_emission_per_unit_denver * np.sum(demand_cfp))
    else:
        transport_emissions = np.dot(transport_co2_array[location_index], demand_bsc) + np.dot(transport_co2_array[location_index], demand_cfp)
        production_emissions = (bsc_emission_per_unit_new * np.sum(demand_bsc)) + (cfp_emission_per_unit_new * np.sum(demand_cfp))

    total_emissions = transport_emissions + production_emissions
    return total_emissions / 1e6  # Convert to millions of kg

# Calculate emissions for all locations
locations = ["Charlotte", "Columbus", "Phoenix", "Denver"]
emissions_results = []

for i, location in enumerate(locations):
    if location == "Denver":
        total_emissions = calculate_total_emissions(i, is_denver=True)
    else:
        total_emissions = calculate_total_emissions(i)
    emissions_results.append([location, total_emissions])

# Format emissions to 2 decimal places
emissions_df["Total CO2 Emissions (Millions kg)"] = emissions_df["Total CO2 Emissions (Millions kg)"].round(2)

# Display results
display(emissions_df)

# Save to CSV if needed
emissions_df.to_csv("co2_emissions_with_denver.csv", index=False)

Location  Total CO2 Emissions (Millions kg)
0  Charlotte                              54.25
1   Columbus                              50.71
2    Phoenix                              92.24
3     Denver                              76.19

### c) Market Coverage Improvement: a. Assess if 2-day delivery to key regions is possible. - Part 1

In [75]:
import numpy as np
import pandas as pd

# Input Data
location_distances = {
    "Charlotte": np.array([634, 725, 755, 1015, 2800]),
    "Columbus": np.array([536, 1030, 360, 1000, 2400]),
    "Phoenix": np.array([2400, 2375, 1750, 1065, 1430]),
    "Denver": np.array([1780, 2065, 1002, 794, 1320])  # Include Denver for comparison
}

# Constants
average_speed = 60  # Average speed in miles per hour

# Function to calculate delivery times in hours
def calculate_delivery_times(distances):
    delivery_times = distances / average_speed  # Time in hours for each region
    return delivery_times

# Analyze each location
results = []
for location, distances in location_distances.items():
    delivery_times = calculate_delivery_times(distances)
    results.append([location] + list(delivery_times))

# Create a DataFrame for output
regions = ["Northeast", "Southeast", "Midwest", "Southwest", "Northwest"]
columns = ["Location"] + [f"Time to {region} (hours)" for region in regions]
results_df = pd.DataFrame(results, columns=columns)

# Round delivery times to 2 decimal places
for column in results_df.columns[1:]:
    results_df[column] = results_df[column].round(2)

# Display results
display(results_df)

# Save to CSV if needed
results_df.to_csv("delivery_times_hours.csv", index=False)


Location  Time to Northeast (hours)  Time to Southeast (hours)  \
0  Charlotte                      10.57                      12.08   
1   Columbus                       8.93                      17.17   
2    Phoenix                      40.00                      39.58   
3     Denver                      29.67                      34.42   

   Time to Midwest (hours)  Time to Southwest (hours)  \
0                    12.58                      16.92   
1                     6.00                      16.67   
2                    29.17                      17.75   
3                    16.70                      13.23   

   Time to Northwest (hours)  
0                      46.67  
1                      40.00  
2                      23.83  
3                      22.00

In [76]:
analysis = """
### Comprehensive Analysis of 2-Day Delivery Feasibility

Based on the calculated delivery times, the feasibility of achieving consistent 2-day delivery (within 48 hours) varies significantly across locations and regions. Below is a detailed analysis for each location:

- **Charlotte**:
    - Charlotte performs well for most regions, with delivery times under 17 hours for the Northeast, Southeast, Midwest, and Southwest.
    - However, delivery to the Northwest is close to the 48-hour threshold at **46.67 hours**. This leaves almost no margin for unforeseen delays such as traffic, weather, or logistical challenges. Serving the Northwest from Denver instead would improve reliability.

- **Columbus**:
    - Columbus provides excellent delivery times to the Midwest (**6.00 hours**) and competitive times to the Northeast (**8.93 hours**).
    - The Northwest is a notable challenge at **40.00 hours**, which is within the 2-day limit but too close for consistent on-time performance. Like Charlotte, this region might be better served from Denver.
    - Delivery to the Southeast (**17.17 hours**) and Southwest (**16.67 hours**) is well within feasible limits.

- **Phoenix**:
    - Phoenix struggles with delivery to the Northeast (**40.00 hours**) and Southeast (**39.58 hours**), making these routes highly susceptible to delays. Additionally, the lack of proximity to these regions makes Phoenix less ideal for serving them regularly.
    - The Southwest (**17.75 hours**) and Northwest (**23.83 hours**) are well within 2-day delivery feasibility. For the Midwest (**29.17 hours**), Phoenix remains a reasonable option.

- **Denver**:
    - Denver generally provides reliable delivery times across all regions, with its central location offering significant advantages.
    - The Northwest is particularly well-served at **22.00 hours**, making Denver the most reliable hub for this region. The Northeast (**29.67 hours**) and Southeast (**34.42 hours**) are within acceptable limits, though not as competitive as other locations.

### Key Observations and Recommendations

1. **Northwest Challenges**:
    - Both Charlotte and Columbus have delivery times to the Northwest that are too close to the 48-hour limit, making them unreliable for consistent 2-day delivery. Denver should serve as the primary hub for this region.

2. **Northeast and Southeast Delivery**:
    - Phoenix struggles significantly with the Northeast and Southeast due to its distance. Alternative hubs such as Charlotte, Columbus, or Denver should handle these regions.

3. **Relay Trucking Considerations**:
    - While relay trucking (using multiple drivers to ensure non-stop travel) could theoretically mitigate delays, this approach may be cost-prohibitive, especially for routes already nearing the 48-hour mark.

4. **Operational Contingencies**:
    - Regions with delivery times exceeding 40 hours should be evaluated for alternative solutions, such as splitting regional coverage between two hubs or leveraging expedited shipping methods.

In conclusion, while all locations can theoretically meet the 2-day delivery window, some routes are too close to the threshold to guarantee consistent reliability. Denver's central location makes it the most robust fallback option for regions with tight delivery constraints, particularly the Northwest.
"""

display(Markdown(analysis))



### Comprehensive Analysis of 2-Day Delivery Feasibility

Based on the calculated delivery times, the feasibility of achieving consistent 2-day delivery (within 48 hours) varies significantly across locations and regions. Below is a detailed analysis for each location:

- **Charlotte**:
    - Charlotte performs well for most regions, with delivery times under 17 hours for the Northeast, Southeast, Midwest, and Southwest.
    - However, delivery to the Northwest is close to the 48-hour threshold at **46.67 hours**. This leaves almost no margin for unforeseen delays such as traffic, weather, or logistical challenges. Serving the Northwest from Denver instead would improve reliability.

- **Columbus**:
    - Columbus provides excellent delivery times to the Midwest (**6.00 hours**) and competitive times to the Northeast (**8.93 hours**).
    - The Northwest is a notable challenge at **40.00 hours**, which is within the 2-day limit but too close for consistent on-time performance. Like Charlotte, this region might be better served from Denver.
    - Delivery to the Southeast (**17.17 hours**) and Southwest (**16.67 hours**) is well within feasible limits.

- **Phoenix**:
    - Phoenix struggles with delivery to the Northeast (**40.00 hours**) and Southeast (**39.58 hours**), making these routes highly susceptible to delays. Additionally, the lack of proximity to these regions makes Phoenix less ideal for serving them regularly.
    - The Southwest (**17.75 hours**) and Northwest (**23.83 hours**) are well within 2-day delivery feasibility. For the Midwest (**29.17 hours**), Phoenix remains a reasonable option.

- **Denver**:
    - Denver generally provides reliable delivery times across all regions, with its central location offering significant advantages.
    - The Northwest is particularly well-served at **22.00 hours**, making Denver the most reliable hub for this region. The Northeast (**29.67 hours**) and Southeast (**34.42 hours**) are within acceptable limits, though not as competitive as other locations.

### Key Observations and Recommendations

1. **Northwest Challenges**:
    - Both Charlotte and Columbus have delivery times to the Northwest that are too close to the 48-hour limit, making them unreliable for consistent 2-day delivery. Denver should serve as the primary hub for this region.

2. **Northeast and Southeast Delivery**:
    - Phoenix struggles significantly with the Northeast and Southeast due to its distance. Alternative hubs such as Charlotte, Columbus, or Denver should handle these regions.

3. **Relay Trucking Considerations**:
    - While relay trucking (using multiple drivers to ensure non-stop travel) could theoretically mitigate delays, this approach may be cost-prohibitive, especially for routes already nearing the 48-hour mark.

4. **Operational Contingencies**:
    - Regions with delivery times exceeding 40 hours should be evaluated for alternative solutions, such as splitting regional coverage between two hubs or leveraging expedited shipping methods.

In conclusion, while all locations can theoretically meet the 2-day delivery window, some routes are too close to the threshold to guarantee consistent reliability. Denver's central location makes it the most robust fallback option for regions with tight delivery constraints, particularly the Northwest.


### C) Evaluate potential cost savings in freight.

In [78]:
transport_costs

{'Charlotte': [3.03, 2.74, 3.92, 5.88, 8.91],
 'Columbus': [2.56, 3.89, 1.87, 5.79, 7.64],
 'Phoenix': [11.46, 8.97, 9.08, 6.17, 4.55],
 'Denver': [8.5, 7.8, 5.2, 4.6, 4.2]}

In [81]:
import numpy as np
import pandas as pd
from IPython.display import display

# Calculate total transportation costs
def calculate_transport_cost(location, transport_costs, demand_bsc, demand_cfp):
    total_cost = (
        np.dot(transport_costs[location], demand_bsc) + 
        np.dot(transport_costs[location], demand_cfp)
    )
    return total_cost

# Calculate costs for all locations
baseline_location = "Denver"
baseline_cost = calculate_transport_cost(baseline_location, transport_costs, demand_bsc, demand_cfp)

results = []
for location in transport_costs.keys():
    total_cost = calculate_transport_cost(location, transport_costs, demand_bsc, demand_cfp)
    cost_savings = baseline_cost - total_cost
    results.append([location, total_cost / 1e6, cost_savings / 1e6])  # Convert to millions

# Create DataFrame
columns = ["Location", "Total Cost (Millions $)", "Cost Savings (Millions $)"]
results_df = pd.DataFrame(results, columns=columns)

# Display results
results_df["Total Cost (Millions $)"] = results_df["Total Cost (Millions $)"].round(2)
results_df["Cost Savings (Millions $)"] = results_df["Cost Savings (Millions $)"].round(2)
display(results_df)

# Save to CSV
results_df.to_csv("freight_cost_savings.csv", index=False)


Location  Total Cost (Millions $)  Cost Savings (Millions $)
0  Charlotte                    56.99                      32.03
1   Columbus                    50.96                      38.06
2    Phoenix                   119.13                     -30.11
3     Denver                    89.02                       0.00

In [83]:
import numpy as np
import pandas as pd
from IPython.display import display

# Transportation cost per unit for each location and region (in $)
transport_costs = {
    "Charlotte": [3.03, 2.74, 3.92, 5.88, 8.91],
    "Columbus": [2.56, 3.89, 1.87, 5.79, 7.64],
    "Phoenix": [11.46, 8.97, 9.08, 6.17, 4.55],
    "Denver": [8.50, 7.80, 5.20, 4.60, 4.20]
}

# Demand per region (units)
demand_bsc = np.array([2500000, 2000000, 1800000, 1200000, 1000000])
demand_cfp = np.array([1500000, 1200000, 1000000, 800000, 500000])

# Region assignments for each split
splits = {
    "Denver + Charlotte": {
        "Denver": [2, 3, 4],  # Midwest, Southwest, Northwest
        "Charlotte": [0, 1]  # Northeast, Southeast
    },
    "Denver + Columbus": {
        "Denver": [3, 4],  # Southwest, Northwest
        "Columbus": [0, 1, 2]  # Northeast, Southeast, Midwest
    },
    "Denver + Phoenix": {
        "Denver": [0, 2, 3],  # Northeast, Midwest, Southwest
        "Phoenix": [1, 4]  # Southeast, Northwest
    }
}

# Function to calculate total cost for a hub serving specific regions
def calculate_cost_for_hub(hub, regions):
    transport_cost = sum(
        np.dot(transport_costs[hub][region], demand_bsc[region]) +
        np.dot(transport_costs[hub][region], demand_cfp[region])
        for region in regions
    )
    return transport_cost

# Calculate costs for each split
baseline_cost = calculate_cost_for_hub("Denver", range(5))  # Denver serves all regions
split_results = []

for split_name, assignment in splits.items():
    total_cost = 0
    for hub, regions in assignment.items():
        total_cost += calculate_cost_for_hub(hub, regions)
    cost_savings = baseline_cost - total_cost
    split_results.append([split_name, total_cost / 1e6, cost_savings / 1e6])  # Convert to millions

# Create a DataFrame for results
columns = ["Split", "Total Cost (Millions $)", "Cost Savings (Millions $)"]
results_df = pd.DataFrame(split_results, columns=columns)

# Format and display results
results_df["Total Cost (Millions $)"] = results_df["Total Cost (Millions $)"].round(2)
results_df["Cost Savings (Millions $)"] = results_df["Cost Savings (Millions $)"].round(2)
display(results_df)

# Save to CSV if needed
results_df.to_csv("freight_cost_savings_splits.csv", index=False)


Split  Total Cost (Millions $)  Cost Savings (Millions $)
0  Denver + Charlotte                    50.95                      38.07
1   Denver + Columbus                    43.42                      45.60
2    Denver + Phoenix                    93.29                      -4.27

In [84]:
import numpy as np
import pandas as pd
from IPython.display import display

# Transportation cost per unit for each location and region (in $)
transport_costs = {
    "Charlotte": [3.03, 2.74, 3.92, 5.88, 8.91],
    "Columbus": [2.56, 3.89, 1.87, 5.79, 7.64],
    "Phoenix": [11.46, 8.97, 9.08, 6.17, 4.55],
    "Denver": [8.50, 7.80, 5.20, 4.60, 4.20]
}

# Demand per region (units)
demand_bsc = np.array([2500000, 2000000, 1800000, 1200000, 1000000])
demand_cfp = np.array([1500000, 1200000, 1000000, 800000, 500000])

# Region assignments for each split
splits = {
    "Denver + Charlotte": {
        "Denver": [2, 3, 4],  # Midwest, Southwest, Northwest
        "Charlotte": [0, 1]  # Northeast, Southeast
    },
    "Denver + Columbus": {
        "Denver": [3, 4],  # Southwest, Northwest
        "Columbus": [0, 1, 2]  # Northeast, Southeast, Midwest
    },
    "Denver + Phoenix": {
        "Denver": [0, 2, 3],  # Northeast, Midwest, Southwest
        "Phoenix": [1, 4]  # Southeast, Northwest
    }
}

# Function to calculate total cost for a hub serving specific regions
def calculate_cost_for_hub(hub, regions):
    transport_cost = sum(
        transport_costs[hub][region] * (demand_bsc[region] + demand_cfp[region])
        for region in regions
    )
    return transport_cost

# Calculate costs for each split
baseline_cost = calculate_cost_for_hub("Denver", range(5))  # Denver serves all regions
split_results = []

for split_name, assignment in splits.items():
    total_cost = 0
    split_details = []
    for hub, regions in assignment.items():
        hub_cost = calculate_cost_for_hub(hub, regions)
        total_cost += hub_cost
        served_regions = ', '.join(["NE", "SE", "MW", "SW", "NW"][r] for r in regions)
        split_details.append(f"{hub} ({served_regions})")
    cost_savings = baseline_cost - total_cost
    split_results.append([split_name + " | " + " & ".join(split_details), total_cost / 1e6, cost_savings / 1e6])

# Create a DataFrame for results
columns = ["Split", "Total Cost (Millions $)", "Cost Savings (Millions $)"]
results_df = pd.DataFrame(split_results, columns=columns)

# Format and display results
results_df["Total Cost (Millions $)"] = results_df["Total Cost (Millions $)"].round(2)
results_df["Cost Savings (Millions $)"] = results_df["Cost Savings (Millions $)"].round(2)
display(results_df)

# Save to CSV if needed
results_df.to_csv("freight_cost_savings_splits.csv", index=False)


Split  Total Cost (Millions $)  \
0  Denver + Charlotte | Denver (MW, SW, NW) & Cha...                    50.95   
1  Denver + Columbus | Denver (SW, NW) & Columbus...                    43.42   
2  Denver + Phoenix | Denver (NE, MW, SW) & Phoen...                    93.29   

   Cost Savings (Millions $)  
0                      38.07  
1                      45.60  
2                      -4.27

### d) Qualitative Factors: Labor availability, environmental regulations, and sustainability incentives

When evaluating potential locations for facility expansion, it's crucial to consider qualitative factors such as labor availability, environmental regulations, and sustainability incentives. Here's an overview of these factors for Charlotte, NC; Columbus, OH; and Phoenix, AZ:

**Labor Availability**

- **Charlotte, NC**: Charlotte boasts a diverse and growing workforce, supported by a range of educational institutions and training programs. The city's economic development initiatives focus on attracting and retaining talent across various industries.

- **Columbus, OH**: Columbus benefits from a young and educated labor pool, partly due to the presence of The Ohio State University. The city emphasizes workforce development to meet the needs of emerging sectors.

- **Phoenix, AZ**: Phoenix has experienced significant population growth, contributing to a large labor force. However, certain specialized skill sets may be in higher demand, necessitating targeted training programs. 

**Environmental Regulations**

- **Charlotte, NC**: The city is committed to environmental sustainability, as evidenced by the Sustainable and Resilient Charlotte by 2050 Resolution. This initiative aims to reduce greenhouse gas emissions and promote sustainable practices. 

- **Columbus, OH**: Columbus has implemented the Sustainable Columbus initiative, focusing on reducing the city's carbon footprint and enhancing environmental quality. The city collaborates with businesses and residents to achieve these goals. 

- **Phoenix, AZ**: Phoenix has established various environmental programs to address issues such as air quality and water conservation. The city encourages sustainable development practices to balance growth with environmental stewardship.

**Sustainability Incentives**

- **Charlotte, NC**: Charlotte offers incentives for sustainable building practices, including policies that promote energy efficiency and the use of renewable energy sources in municipal facilities. 

- **Columbus, OH**: The city provides programs like Clean Energy Columbus, which connects residents and businesses with renewable energy options. Additionally, initiatives such as GreenSpot encourage organizations to adopt green practices. 

- **Phoenix, AZ**: Phoenix offers various sustainability programs and may provide incentives for businesses that implement environmentally friendly practices, though specific programs would need to be confirmed through local resources.

In summary, each city presents unique advantages regarding labor availability, environmental regulations, and sustainability incentives. Charlotte and Columbus have explicit sustainability initiatives and incentives, while Phoenix's offerings may require further investigation to identify specific programs. 

## e) Recommendation: Which location (or combination) best meets GreenPack’s goals (profit, speed, sustainability)

**Recommendation**:

Based on the output provided, the combination of **Denver + Columbus** emerges as the best choice for GreenPack’s expansion, considering profitability, speed, and sustainability. Here's how it meets the goals:

1. **Profitability**:
   - **Total Cost**: Denver + Columbus has the **lowest total cost** at $43.42M, which is significantly lower than the other options.
  
     
   - **Cost Savings**: This combination yields the **highest cost savings** of $45.60M compared to operating from Denver alone.

2. **Speed**:
   - Columbus’s **central location** provides efficient access to multiple regions, including the Northeast, Midwest, and Southeast, enabling faster deliveries.
   - By splitting operations, Denver can continue serving the Southwest and Northwest efficiently, while Columbus reduces the burden for regions farther east.

3. **Sustainability**:
   - Shorter transportation distances from Columbus to high-demand areas (e.g., Northeast and Southeast) **reduce transportation emissions**, contributing to GreenPack’s sustainability goals.
   - Columbus has access to renewable energy initiatives (e.g., Clean Energy Columbus) and fosters green business practices through programs like GreenSpot.

### Summary of Other Options:

- **Denver + Charlotte**:
   - **Total Cost**: $50.95M, higher than Denver + Columbus.
   - While Charlotte is closer to the Southeast and Northeast, its higher fixed costs and environmental regulation complexity make it less cost-effective.

- **Denver + Phoenix**:
   - **Total Cost**: $93.29M, making it the **most expensive option**.
   - Phoenix's distance from high-demand regions (e.g., Northeast and Southeast) results in **higher transportation costs** and emissions, reducing its alignment with profitability and sustainability goals.

### Conclusion:
The combination of **Denver + Columbus** is the optimal recommendation for GreenPack, offering the best balance of cost efficiency, delivery speed, and sustainability. It strategically splits operations to minimize transportation costs and emissions while leveraging Columbus’s central location and green initiatives.

# Part 3

In [107]:
from IPython.display import display, Markdown

# Corrected LaTeX write-up
write_up = r"""
## Optimization Model Explanation

### Objective Function
The goal is to minimize the total cost \( Z \), which comprises the following components:

1. **Fixed Costs**:
   $$ 
   \sum_{i \in F} F_i \cdot y_i 
   $$
   - \( F_i \): Fixed cost of opening facility \( i \).
   - \( y_i \): Binary decision variable, where \( y_i = 1 \) if facility \( i \) is open, and \( y_i = 0 \) otherwise.

2. **Production Costs**:
   $$ 
   \sum_{i \in F} \sum_{j \in R} \left( P_{BSC,i} \cdot x_{BSC,ij} + P_{CFP,i} \cdot x_{CFP,ij} \right) 
   $$
   - \( P_{BSC,i}, P_{CFP,i} \): Production cost per unit for BSC and CFP at facility \( i \).
   - \( x_{BSC,ij}, x_{CFP,ij} \): Number of units of BSC and CFP shipped from facility \( i \) to region \( j \).

3. **Transportation Costs**:
   $$ 
   \sum_{i \in F} \sum_{j \in R} \left( T_{ij} \cdot (x_{BSC,ij} + x_{CFP,ij}) \right) 
   $$
   - \( T_{ij} \): Transportation cost per unit from facility \( i \) to region \( j \).

4. **Emissions Costs**:
   $$ 
   \text{CarbonPrice} \cdot \left( 
   \sum_{i \in F} \sum_{j \in R} \left( E_{P,BSC,i} \cdot x_{BSC,ij} + E_{P,CFP,i} \cdot x_{CFP,ij} \right) + 
   \sum_{i \in F} \sum_{j \in R} \left( E_{T,ij} \cdot (x_{BSC,ij} + x_{CFP,ij}) \right)
   \right) 
   $$
   - \( \text{CarbonPrice} \): Internal carbon price per kilogram of CO\(_2\).
   - \( E_{P,BSC,i}, E_{P,CFP,i} \): Production emissions (kg CO\(_2\) per unit of BSC and CFP) at facility \( i \).
   - \( E_{T,ij} \): Transportation emissions (kg CO\(_2\) per unit shipped from facility \( i \) to region \( j \).

### Constraints
The objective function is subject to several constraints, including:

1. **Demand Satisfaction**:
   $$
   \sum_{i \in F} x_{BSC,ij} \geq D_{BSC,j}, \quad \forall j \in R 
   $$
   $$
   \sum_{i \in F} x_{CFP,ij} \geq D_{CFP,j}, \quad \forall j \in R 
   $$
   Ensures that the total units shipped meet or exceed regional demands.

2. **Capacity Limits**:
   $$
   \sum_{j \in R} x_{BSC,ij} \leq C_{BSC,i} \cdot y_i, \quad \forall i \in F 
   $$
   $$
   \sum_{j \in R} x_{CFP,ij} \leq C_{CFP,i} \cdot y_i, \quad \forall i \in F 
   $$
   Ensures that production does not exceed the available capacity at any facility.

3. **Facility Opening**:
   $$
   y_i \in \{0, 1\}, \quad \forall i \in F 
   $$

4. **Non-Negativity**:
   $$
   x_{BSC,ij}, x_{CFP,ij} \geq 0, \quad \forall i \in F, j \in R 
   $$

This function ensures a balance between cost minimization and sustainability, aligning with GreenPack's strategic goals.
"""

# Display the explanation as Markdown
display(Markdown(write_up))



## Optimization Model Explanation

### Objective Function
The goal is to minimize the total cost \( Z \), which comprises the following components:

1. **Fixed Costs**:
   $$ 
   \sum_{i \in F} F_i \cdot y_i 
   $$
   - \( F_i \): Fixed cost of opening facility \( i \).
   - \( y_i \): Binary decision variable, where \( y_i = 1 \) if facility \( i \) is open, and \( y_i = 0 \) otherwise.

2. **Production Costs**:
   $$ 
   \sum_{i \in F} \sum_{j \in R} \left( P_{BSC,i} \cdot x_{BSC,ij} + P_{CFP,i} \cdot x_{CFP,ij} \right) 
   $$
   - \( P_{BSC,i}, P_{CFP,i} \): Production cost per unit for BSC and CFP at facility \( i \).
   - \( x_{BSC,ij}, x_{CFP,ij} \): Number of units of BSC and CFP shipped from facility \( i \) to region \( j \).

3. **Transportation Costs**:
   $$ 
   \sum_{i \in F} \sum_{j \in R} \left( T_{ij} \cdot (x_{BSC,ij} + x_{CFP,ij}) \right) 
   $$
   - \( T_{ij} \): Transportation cost per unit from facility \( i \) to region \( j \).

4. **Emissions Costs**:
   $$ 
   \text{CarbonPrice} \cdot \left( 
   \sum_{i \in F} \sum_{j \in R} \left( E_{P,BSC,i} \cdot x_{BSC,ij} + E_{P,CFP,i} \cdot x_{CFP,ij} \right) + 
   \sum_{i \in F} \sum_{j \in R} \left( E_{T,ij} \cdot (x_{BSC,ij} + x_{CFP,ij}) \right)
   \right) 
   $$
   - \( \text{CarbonPrice} \): Internal carbon price per kilogram of CO\(_2\).
   - \( E_{P,BSC,i}, E_{P,CFP,i} \): Production emissions (kg CO\(_2\) per unit of BSC and CFP) at facility \( i \).
   - \( E_{T,ij} \): Transportation emissions (kg CO\(_2\) per unit shipped from facility \( i \) to region \( j \).

### Constraints
The objective function is subject to several constraints, including:

1. **Demand Satisfaction**:
   $$
   \sum_{i \in F} x_{BSC,ij} \geq D_{BSC,j}, \quad \forall j \in R 
   $$
   $$
   \sum_{i \in F} x_{CFP,ij} \geq D_{CFP,j}, \quad \forall j \in R 
   $$
   Ensures that the total units shipped meet or exceed regional demands.

2. **Capacity Limits**:
   $$
   \sum_{j \in R} x_{BSC,ij} \leq C_{BSC,i} \cdot y_i, \quad \forall i \in F 
   $$
   $$
   \sum_{j \in R} x_{CFP,ij} \leq C_{CFP,i} \cdot y_i, \quad \forall i \in F 
   $$
   Ensures that production does not exceed the available capacity at any facility.

3. **Facility Opening**:
   $$
   y_i \in \{0, 1\}, \quad \forall i \in F 
   $$

4. **Non-Negativity**:
   $$
   x_{BSC,ij}, x_{CFP,ij} \geq 0, \quad \forall i \in F, j \in R 
   $$

This function ensures a balance between cost minimization and sustainability, aligning with GreenPack's strategic goals.


In [108]:
from IPython.display import display, Math

# LaTeX code for the optimization function
latex_code = r"""
\text{Minimize: } Z = 
\underbrace{\sum_{i \in F} F_i \cdot y_i}_{\text{Fixed Costs}} +
\underbrace{\sum_{i \in F} \sum_{j \in R} \left( P_{BSC,i} \cdot x_{BSC,ij} + P_{CFP,i} \cdot x_{CFP,ij} \right)}_{\text{Production Costs}} +
\underbrace{\sum_{i \in F} \sum_{j \in R} \left( T_{ij} \cdot (x_{BSC,ij} + x_{CFP,ij}) \right)}_{\text{Transportation Costs}} +
\underbrace{\text{CarbonPrice} \cdot \left( 
\sum_{i \in F} \sum_{j \in R} \left( E_{P,BSC,i} \cdot x_{BSC,ij} + E_{P,CFP,i} \cdot x_{CFP,ij} \right) + 
\sum_{i \in F} \sum_{j \in R} \left( E_{T,ij} \cdot (x_{BSC,ij} + x_{CFP,ij}) \right)
\right)}_{\text{Emissions Costs}}
"""

# Display LaTeX in Jupyter Notebook
display(Math(latex_code))


<IPython.core.display.Math object>

In [121]:
from pulp import LpProblem, LpMinimize, LpVariable, lpSum, LpBinary

# 1) Define sets
facilities = ['Denver', 'Charlotte', 'Columbus', 'Phoenix']
regions = ['Northeast', 'Southeast', 'Midwest', 'Southwest', 'Northwest']

# 2) Parameters (from your code snippet or re-arranged as dicts)

# Demand
demand_BSC = {
    'Northeast': 2_500_000,
    'Southeast': 2_000_000,
    'Midwest':   1_800_000,
    'Southwest': 1_200_000,
    'Northwest': 1_000_000
}
demand_CFP = {
    'Northeast': 1_500_000,
    'Southeast': 1_200_000,
    'Midwest':   1_000_000,
    'Southwest':   800_000,
    'Northwest':   500_000
}

# Facility capacities (BSC & CFP)
capacity_BSC = {
    'Denver':    8_000_000,
    'Charlotte': 4_000_000,  # If built
    'Columbus':  4_000_000,  # If built
    'Phoenix':   4_000_000   # If built
}
capacity_CFP = {
    'Denver':    5_000_000,
    'Charlotte': 3_000_000,
    'Columbus':  3_000_000,
    'Phoenix':   3_000_000
}

# Fixed costs for each facility's lines.
# If you only want "one binary open cost" per facility (both lines together),
# then you can combine them. Here we'll separate for clarity:
fixed_cost_BSC = {
    'Denver':    12_000_000,  # Already open, so you might skip or treat differently
    'Charlotte':  7_000_000,
    'Columbus':   7_000_000,
    'Phoenix':    7_000_000
}
fixed_cost_CFP = {
    'Denver':     8_000_000, 
    'Charlotte':  5_000_000,
    'Columbus':   5_000_000,
    'Phoenix':    5_000_000
}

# Variable production cost
prod_cost_BSC = {
    'Denver':    15,
    'Charlotte': 16,
    'Columbus':  16,
    'Phoenix':   16
}
prod_cost_CFP = {
    'Denver':    8,
    'Charlotte': 8.5,
    'Columbus':  8.5,
    'Phoenix':   8.5
}

# Production emissions (kg/unit)
prod_em_BSC = {
    'Denver':    2.0,
    'Charlotte': 1.8,
    'Columbus':  1.8,
    'Phoenix':   1.8
}
prod_em_CFP = {
    'Denver':    1.0,
    'Charlotte': 0.9,
    'Columbus':  0.9,
    'Phoenix':   0.9
}

# Transport costs ($/unit) and transport emissions (kg/unit)
transport_cost = {
    'Denver':    [8.5, 7.8, 5.2, 4.6, 4.2],
    'Charlotte': [3.027528, 2.738499, 3.918164, 5.880353, 8.909091],
    'Columbus':  [2.559551, 3.890557, 1.868263, 5.793451, 7.636364],
    'Phoenix':   [11.460674, 8.970944, 9.081836, 6.170025, 4.55]
}
transport_em = {
    'Denver':    [5.2, 4.8, 3.1, 2.8, 2.5],
    'Charlotte': [1.852135, 1.685230, 2.335828, 3.579345, 5.303030],
    'Columbus':  [1.565843, 2.394189, 1.113772, 3.526448, 4.545455],
    'Phoenix':   [7.011236, 5.520581, 5.414172, 3.755668, 2.708333]
}
# We'll map these by region index, so let's keep a region order list:
region_index = { 'Northeast': 0, 'Southeast': 1, 'Midwest': 2, 'Southwest': 3, 'Northwest': 4 }

# Carbon price
carbon_price = 0.10

# 3) Create the model
model = LpProblem("GreenPack_Optimization", LpMinimize)

# 4) Decision variables
#   xBSC[i,j] = quantity of BSC shipped from facility i to region j
#   xCFP[i,j] = quantity of CFP shipped from facility i to region j
#   yBSC[i]   = 1 if we open BSC line at facility i, 0 otherwise
#   yCFP[i]   = 1 if we open CFP line at facility i, 0 otherwise
#
#   If Denver is already open for both lines, we can fix them to 1 (or skip them).
xBSC = {}
xCFP = {}
for i in facilities:
    for j in regions:
        xBSC[(i,j)] = LpVariable(f"xBSC_{i}_{j}", lowBound=0, cat="Continuous")
        xCFP[(i,j)] = LpVariable(f"xCFP_{i}_{j}", lowBound=0, cat="Continuous")

# For new facilities, we have binary open/close. For Denver, we can either fix to 1 or just keep it as 1 in constraints.
yBSC = {}
yCFP = {}
for i in facilities:
    if i == "Denver":
        # If Denver is definitely open for both lines:
        yBSC[i] = LpVariable(f"yBSC_{i}", lowBound=1, upBound=1, cat="Continuous")  # fixed to 1
        yCFP[i] = LpVariable(f"yCFP_{i}", lowBound=1, upBound=1, cat="Continuous")  # fixed to 1
    else:
        yBSC[i] = LpVariable(f"yBSC_{i}", cat=LpBinary)
        yCFP[i] = LpVariable(f"yCFP_{i}", cat=LpBinary)

# 5) Objective function
#    Z = sum_i[ fixed(BSC,i)*yBSC[i] + fixed(CFP,i)*yCFP[i] ]
#         + sum_{i,j}[ (prod_cost_BSC[i] + transport_cost[i,j] + carbon_price * (prod_em_BSC[i]+transport_em[i,j])) * xBSC[i,j] ]
#         + sum_{i,j}[ (prod_cost_CFP[i] + transport_cost[i,j] + carbon_price * (prod_em_CFP[i]+transport_em[i,j])) * xCFP[i,j] ]
model += (
    lpSum(fixed_cost_BSC[i] * yBSC[i] for i in facilities) +
    lpSum(fixed_cost_CFP[i] * yCFP[i] for i in facilities) +
    lpSum(
        (prod_cost_BSC[i] 
         + transport_cost[i][region_index[j]] 
         + carbon_price * (prod_em_BSC[i] + transport_em[i][region_index[j]])) 
        * xBSC[(i,j)]
        for i in facilities for j in regions
    ) +
    lpSum(
        (prod_cost_CFP[i] 
         + transport_cost[i][region_index[j]] 
         + carbon_price * (prod_em_CFP[i] + transport_em[i][region_index[j]])) 
        * xCFP[(i,j)]
        for i in facilities for j in regions
    )
), "Total_Cost"

# 6) Constraints

# Demand must be met in each region for BSC and CFP
for j in regions:
    model += lpSum( xBSC[(i,j)] for i in facilities ) == demand_BSC[j], f"BSC_demand_{j}"
    model += lpSum( xCFP[(i,j)] for i in facilities ) == demand_CFP[j], f"CFP_demand_{j}"

# Capacity constraints for each facility
for i in facilities:
    # BSC capacity
    model += lpSum(xBSC[(i,j)] for j in regions) <= capacity_BSC[i] * yBSC[i], f"BSC_capacity_{i}"
    # CFP capacity
    model += lpSum(xCFP[(i,j)] for j in regions) <= capacity_CFP[i] * yCFP[i], f"CFP_capacity_{i}"

# If you want to force that a facility is either open for both or not at all, you can do:
# yBSC[i] == yCFP[i], but that is optional. 
# Or you can keep them separate to allow partial expansions.

# 7) Solve the model
model.solve()

print("Status:", model.status)
print("Objective value (Total Cost):", model.objective.value())

# 8) Print out the decisions
for i in facilities:
    print(f"\nFacility: {i}")
    print(f"  yBSC[{i}] = {yBSC[i].varValue}")
    print(f"  yCFP[{i}] = {yCFP[i].varValue}")
    for j in regions:
        flow_bsc = xBSC[(i,j)].varValue
        flow_cfp = xCFP[(i,j)].varValue
        if flow_bsc > 1e-5 or flow_cfp > 1e-5:
            print(f"    -> Region {j}: BSC={flow_bsc:,.0f}, CFP={flow_cfp:,.0f}")


Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /Users/danya/anaconda/anaconda3/lib/python3.12/site-packages/pulp/solverdir/cbc/osx/64/cbc /var/folders/ph/vn1yxmc909ld_37rxq3d0g1h0000gn/T/3a02f842e9864bb1b6b7be85a73b4ed6-pulp.mps -timeMode elapsed -branch -printingOptions all -solution /var/folders/ph/vn1yxmc909ld_37rxq3d0g1h0000gn/T/3a02f842e9864bb1b6b7be85a73b4ed6-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 23 COLUMNS
At line 172 RHS
At line 191 BOUNDS
At line 200 ENDATA
Problem MODEL has 18 rows, 48 columns and 88 elements
Coin0008I MODEL read with 0 errors
Option for timeMode changed from cpu to elapsed
Continuous objective value is 2.57173e+08 - 0.00 seconds
Cgl0004I processed model has 17 rows, 46 columns (6 integer (6 of which binary)) and 81 elements
Cbc0038I Initial state - 3 integers unsatisfied sum - 0.991667
Cbc0038I Pass   1: suminf.    0.00000 (0) obj. 2.65685e+08 iterations 4
Cbc0038I Soluti

In [122]:
import pandas as pd

# Print overall objective value nicely
print(f"Objective value (Total Cost): ${model.objective.value():,.2f}\n")

# Summarize which facilities and lines are open
print("=== Facility/Line Open Status ===")
for i in facilities:
    # Convert floats to int for 0/1 display
    bsc_open = int(yBSC[i].varValue)
    cfp_open = int(yCFP[i].varValue)
    print(f"Facility: {i:<9} | BSC Open? {bsc_open} | CFP Open? {cfp_open}")
print()

# Collect shipping results into a DataFrame
solution_rows = []
for i in facilities:
    for j in regions:
        bsc_flow = xBSC[(i,j)].varValue
        cfp_flow = xCFP[(i,j)].varValue
        # Only record rows where there is any flow > 0
        if (bsc_flow > 1e-6) or (cfp_flow > 1e-6):
            solution_rows.append({
                "Facility": i,
                "Region": j,
                "BSC Shipped": bsc_flow,
                "CFP Shipped": cfp_flow
            })

solution_df = pd.DataFrame(solution_rows)
solution_df = solution_df[["Facility", "Region", "BSC Shipped", "CFP Shipped"]]  # reorder columns

# Print a neat table
print("=== Shipment Plan ===")
display(solution_df.round({"BSC Shipped": 0, "CFP Shipped": 0}))


Objective value (Total Cost): $262,505,779.16

=== Facility/Line Open Status ===
Facility: Denver    | BSC Open? 1 | CFP Open? 1
Facility: Charlotte | BSC Open? 1 | CFP Open? 1
Facility: Columbus  | BSC Open? 1 | CFP Open? 0
Facility: Phoenix   | BSC Open? 0 | CFP Open? 0

=== Shipment Plan ===


Facility     Region  BSC Shipped  CFP Shipped
0     Denver    Midwest          0.0     700000.0
1     Denver  Southwest    1200000.0     800000.0
2     Denver  Northwest    1000000.0     500000.0
3  Charlotte  Northeast     300000.0    1500000.0
4  Charlotte  Southeast    2000000.0    1200000.0
5  Charlotte    Midwest          0.0     300000.0
6   Columbus  Northeast    2200000.0          0.0
7   Columbus    Midwest    1800000.0          0.0

# Part 4

Our optimized supply chain network carefully balances cost-efficiency, carbon footprint, and regional advantages. By strategically combining Denver with one or more additional facilities, we reduce transportation emissions and operating costs while enhancing alignment with GreenPack's sustainability goals.

Carbon Footprint: The optimization results show that adding Charlotte or Columbus significantly reduces transportation emissions due to their proximity to high-demand regions like the Southeast and Midwest. For example, allocating Southeast demand to Charlotte reduces emissions by approximately X% compared to serving the region solely from Denver. Conversely, Phoenix contributes higher emissions for East Coast regions but excels in serving Southwest demand with minimal transport emissions.

Water Usage: Water scarcity is a key concern in Phoenix, which would require investments in water-efficient technologies or recycled water systems to mitigate environmental impact. Columbus and Charlotte face fewer water-related challenges, making them better suited for water-intensive production.

Renewable Energy Adoption: Charlotte and Phoenix stand out for their renewable energy potential. North Carolina’s state green incentives and Arizona’s year-round solar capacity provide opportunities for renewable-powered operations. However, Columbus has limited access to renewables, which could increase the carbon intensity of operations unless investments are made in renewable sourcing.

Waste Reduction: Our optimized network ensures efficient facility utilization, minimizing waste from overproduction or idle capacity. Centralizing operations in Columbus or decentralizing across Denver and Charlotte achieves balanced facility utilization while maintaining flexibility for future growth.

Local Community Impact: Charlotte and Phoenix offer strong advantages for local community impact. Charlotte’s skilled workforce and green job potential align with GreenPack’s sustainability goals. Phoenix, with its growing tech sector and university R&D partnerships, fosters innovation in sustainable practices. Columbus, while cost-effective, may require additional investment in workforce training and renewable energy infrastructure to meet GreenPack’s long-term goals.

Regional Trade-offs: Charlotte excels in serving East Coast demand with low transport emissions and access to green incentives but comes with higher labor costs and stricter environmental regulations. Columbus offers a central location with lower costs and strong logistics infrastructure, though its limited renewable options may hinder sustainability progress. Phoenix, while far from East Coast demand, leverages its solar potential and growing tech workforce to support innovative, sustainable operations.

In conclusion, the optimized network demonstrates that adding Charlotte or Columbus to the existing Denver facility offers the best balance of cost, sustainability, and community impact. While Phoenix is less ideal for nationwide demand due to distance, it could serve as a future hub for sustainable innovation in the Southwest. By leveraging regional advantages and addressing challenges such as water scarcity or renewables, GreenPack can build a resilient and sustainable supply chain.

\
